### DISCUSSION TITLE
bug - undocumented that http_method_names requires lowercase

### DISCUSSION THREAD

**Discussion by bsolomon1124** [encode/django-rest-framework#8118] [Potential Issue]
Works:

```python
class UserViewSet(viewsets.ModelViewSet):
    http_method_names = [i.lower() for i in permissions.SAFE_METHODS]
```

Breaks:

```python
class UserViewSet(viewsets.ModelViewSet):
    http_method_names = permissions.SAFE_METHODS
```

The lowercase version shows `Allow: GET, HEAD, OPTIONS` in the browser UI, while leaving SAFE_METHODS as-is with zero other changes brings back `"detail": "Method \"GET\" not allowed."`.

I have confirmed this is not due to `list` v `tuple`, solely due to case.

**Comment by bsolomon1124** [comment]
I believe the faulty comparison may lie here:

https://github.com/encode/django-rest-framework/blob/71e6c30034a1dd35a39ca74f86c371713e762c79/rest_framework/views.py#L499

Where the LHS is lowercased therefore assuming `self.http_method_names` is also lower.

**Comment by lovelydinosaur** [comment]
I think that using uppercased names for `http_method_names` would result in slightly broken behaviour anyways...

Taking a look here: https://github.com/django/django/blob/main/django/views/generic/base.py#L117 - Django expects to be able to map the name directly onto the counterpart method, when determining which methods to add to the `Allow` header.

I'm not sure if accepting #8107 would necessarily be a great idea, because I think it'd mean that the REST framework dispatch would then start working if uppercased method names are used in `http_method_names`, but the `Allow` headers that Django generates for those views would be broken.

(I think?)
